cql3/restrictions/statement_restrictions.cc

```diff
@@ -1068,6 +1068,110 @@ std::vector<query::clustering_range> get_single_column_clustering_bounds(
+    token_column_bigint.type = long_type;
+    expression new_token_restrictions = replace_column_def(token_restriction, &token_column_bigint);
+
+    std::variant<value_list, nonwrapping_range<managed_bytes>> values =
+        possible_lhs_values(&token_column_bigint, new_token_restrictions, options);
+
+    return std::visit(overloaded_functor {
+        [](const value_list& list) {
+            if (list.empty()) { // Impossible condition -- no rows can possibly match.
+                return std::vector<query::clustering_range>();
+            }
+
+            std::vector<query::clustering_range> ck_ranges;
+            ck_ranges.reserve(list.size());
+
+            for (auto&& value : list) {
+                ck_ranges.emplace_back(query::clustering_range::make_singular(std::vector<managed_bytes>{value}));
+            }
+
+            return ck_ranges;
+        },
+        [](const nonwrapping_interval<managed_bytes>& range) {
+            auto int64_from_be_bytes = [](const managed_bytes& int_bytes) -> int64_t {
+                if (int_bytes.size() != 8) {
+                    throw std::runtime_error("token restriction value should be 8 bytes");
+                }
+
+                return read_be<int64_t>((const char*)&int_bytes[0]);
+            };
+
+            auto int64_to_be_bytes = [](int64_t int_val) -> managed_bytes {
+                managed_bytes int_bytes(managed_bytes::initialized_later{}, 8);
+                write_be((char*)&int_bytes[0], int_val);
+                return int_bytes;
+            };
+
+            int64_t token_low = std::numeric_limits<int64_t>::min();
+            int64_t token_high = std::numeric_limits<int64_t>::max();
+            bool low_inclusive = true, high_inclusive = true;
+
+            const std::optional<interval_bound<managed_bytes>>& start = range.start();
+            const std::optional<interval_bound<managed_bytes>>& end = range.end();
+
+            if (start.has_value()) {
+                token_low = int64_from_be_bytes(start->value());
+                low_inclusive = start->is_inclusive();
+            }
+
+            if (end.has_value()) {
+                token_high = int64_from_be_bytes(end->value());
+                high_inclusive = end->is_inclusive();
+            }
+
+            query::clustering_range::bound lower_bound(std::vector{int64_to_be_bytes(token_low)}, low_inclusive);
+            query::clustering_range::bound upper_bound(std::vector{int64_to_be_bytes(token_high)}, high_inclusive);
+
+            std::vector<query::clustering_range> ck_ranges;
+
+            // Blob encoded tokens are sorted like this:
+            // 0, 1, 2, 3, 4, 5, ..., bigint_max, bigint_min, ...., -5, -4, -3, -2, -1
+            // This means that in cases where low >= 0 or high < 0 we can simply use the whole range.
+            // In other cases we have to take two ranges: (low, -1) and (0, high).
+            if (token_low >= 0 || token_high < 0) {
```

Comment:
What if both of these are true (ie, an impossible token restriction)?  I doubt the next line will work correctly; see how this code filters out impossible ranges:
https://github.com/scylladb/scylla/blob/7261c2c93e633a5df2772cc1313da59712ba74d4/cql3/restrictions/single_column_primary_key_restrictions.hh#L379-L385

Reply:
The range is from `possible_lhs_values`, so it should be correct probably (?)

Reply:
That code I quoted also gets wrapping_bounds from possible_lhs_values (via compute_bounds), and it still needs to filter out impossible ranges.

Reply:
I added
```c++
if (token_high < token_low) {
        // Impossible range, return empty clustering ranges
        return ck_ranges;
   }
```
To fix that

Reply:
Sounds good.